storage-proofs/src/merkle/proof.rs

```diff
@@ -61,6 +61,10 @@ pub trait MerkleProofTrait:
     }
     fn verify(&self) -> bool;
 
+    /// Validates the MerkleProof and that it corresponds to the supplied node.
+    ///
+    /// TODO: audit performance and usage in case verification is
```

Comment:
Not just performance. We should understand how all these methods are used and refactor such that there are clear semantics and have each one only do the job required of it. Although the duplication is wasteful, I'm more concerned about the effect of accumulating slop.